pytest/lib/proxy.py

```diff
@@ -212,6 +212,9 @@ async def bridge(reader, writer, handler_fn, global_stopped, local_stopped,
         while 0 == global_stopped.value and 0 >= local_stopped.value and 0 == error.value and 0 == bridge_stopped[
                 0]:
             header = await reader.read(4)
+            # It's possible that we will not read whole header at once, which can use a crash.
+            while len(header) < 4:
+                header += await reader.read(4 - len(header))
```

Comment:
Perhaps define a helper function:

    async def _read_exact(reader, length):
        """Reads exactly `length` bytes from reader."""
        data = b''
        while True:
            buf = await reader.read(length - len(data))
            if not buf:
                raise RuntimeError('Endpoint closed')
            data += buf
            if len(data) >= length:
                return data

Though I’m not sure whether closed connection is handled correctly with this.

Reply:
I'm  not sure either. Anyway, thanks! I think it's a good idea to use the same reader for header and body.